File: Swift/Sources/CalculationParameters.swift
//
//  CalculationParameters.swift
//  Adhan
//
//  Copyright © 2018 Batoul Apps. All rights reserved.
//

import Foundation

/* All customizable parameters for calculating prayer times */
public struct CalculationParameters {
    public var method: CalculationMethod = .other
    public var fajrAngle: Double
    public var ishaAngle: Double
    public var ishaInterval: Int = 0
    public var madhab: Madhab = .shafi
    public var highLatitudeRule: HighLatitudeRule = .middleOfTheNight
    public var adjustments: PrayerAdjustments = PrayerAdjustments()

    init(fajrAngle: Double, ishaAngle: Double) {
        self.fajrAngle = fajrAngle
        self.ishaAngle = ishaAngle
    }

    init(fajrAngle: Double, ishaInterval: Int) {
        self.init(fajrAngle: fajrAngle, ishaAngle: 0)
        self.ishaInterval = ishaInterval
    }

    init(fajrAngle: Double, ishaAngle: Double, method: CalculationMethod) {
        self.init(fajrAngle: fajrAngle, ishaAngle: ishaAngle)
        self.method = method
    }

    init(fajrAngle: Double, ishaInterval: Int, method: CalculationMethod) {
        self.init(fajrAngle: fajrAngle, ishaInterval: ishaInterval)
        self.method = method
    }

    func nightPortions() -> (fajr: Double, isha: Double) {
        switch self.highLatitudeRule {
        case .middleOfTheNight:
            return (1/2, 1/2)
        case .seventhOfTheNight:
            return (1/7, 1/7)
        case .twilightAngle:
            return (self.fajrAngle / 60, self.ishaAngle / 60)
        }
    }
}

File: java/adhan/src/main/java/com/batoulapps/adhan/CalculationParameters.java
package com.batoulapps.adhan;

/**
 * Parameters used for PrayerTime calculation customization
 *
 * Note that, for many cases, you can use {@link CalculationMethod#getParameters()} to get a
 * pre-computed set of calculation parameters depending on one of the available
 * {@link CalculationMethod}.
 */
public class CalculationParameters {

  /**
   * The method used to do the calculation
   */
  public CalculationMethod method = CalculationMethod.OTHER;

  /**
   * The angle of the sun used to calculate fajr
   */
  public double fajrAngle;

  /**
   * The angle of the sun used to calculate isha
   */
  public double ishaAngle;

  /**
   * Minutes after Maghrib (if set, the time for Isha will be Maghrib plus IshaInterval)
   */
  public int ishaInterval;

  /**
   * The madhab used to calculate Asr
   */
  public Madhab madhab = Madhab.SHAFI;

  /**
   * Rules for placing bounds on Fajr and Isha for high latitude areas
   */
  public HighLatitudeRule highLatitudeRule = HighLatitudeRule.MIDDLE_OF_THE_NIGHT;

  /**
   * Used to optionally add or subtract a set amount of time from each prayer time
   */
  public PrayerAdjustments adjustments = new PrayerAdjustments();

  /**
   * Generate CalculationParameters from angles
   * @param fajrAngle the angle for calculating fajr
   * @param ishaAngle the angle for calculating isha
   */
  public CalculationParameters(double fajrAngle, double ishaAngle) {
    this.fajrAngle = fajrAngle;
    this.ishaAngle = ishaAngle;
  }

  /**
   * Generate CalculationParameters from fajr angle and isha interval
   * @param fajrAngle the angle for calculating fajr
   * @param ishaInterval the amount of time after maghrib to have isha
   */
  public CalculationParameters(double fajrAngle, int ishaInterval) {
    this(fajrAngle, 0.0);
    this.ishaInterval = ishaInterval;
  }

  /**
   * Generate CalculationParameters from angles and a calculation method
   * @param fajrAngle the angle for calculating fajr
   * @param ishaAngle the angle for calculating isha
   * @param method the calculation method to use
   */
  public CalculationParameters(double fajrAngle, double ishaAngle, CalculationMethod method) {
    this(fajrAngle, ishaAngle);
    this.method = method;
  }

  /**
   * Generate CalculationParameters from fajr angle, isha interval, and calculation method
   * @param fajrAngle the angle for calculating fajr
   * @param ishaInterval the amount of time after maghrib to have isha
   * @param method the calculation method to use
   */
  public CalculationParameters(double fajrAngle, int ishaInterval, CalculationMethod method) {
    this(fajrAngle, ishaInterval);
    this.method = method;
  }

  static class NightPortions {
    final double fajr;
    final double isha;

    private NightPortions(double fajr, double isha) {
      this.fajr = fajr;
      this.isha = isha;
    }
  }

  NightPortions nightPortions() {
    switch (this.highLatitudeRule) {
      case MIDDLE_OF_THE_NIGHT: {
        return new NightPortions(1.0 / 2.0, 1.0 / 2.0);
      }
      case SEVENTH_OF_THE_NIGHT: {
        return new NightPortions(1.0 / 7.0, 1.0 / 7.0);
      }
      case TWILIGHT_ANGLE: {
        return new NightPortions(this.fajrAngle / 60.0, this.ishaAngle / 60.0);
      }
      default: {
        throw new IllegalArgumentException("Invalid high latitude rule");
      }
    }
  }
}
